src/pint/models/astrometry.py

```diff
@@ -279,16 +273,17 @@ def validate(self):
         for p in ("RAJ", "DECJ"):
             if getattr(self, p).value is None:
                 raise MissingParameter("Astrometry", p)
+        if (self.PMRA.quantity is None) != (self.PMDEC.quantity is None):
+            raise ValueError("PMELONG and PMELAT must either both be set or neither.")
```

Comment:
The error message is a mistake. My reasoning is that if someone has set PMRA but not PMDEC then there's some error - perhaps they just forgot to include it, but assuming zero is unlikely to be the right answer. I'd rather encounter an exception than a wrong assumption. And it's not like it's hard to just add PMDEC to your par file.

Reply:
Hmmm.  Not sure I agree with that being an error.   Based on where your source is located in the sky, you can easily have 10-20x better precision for one of the PMs than the other, simply because we get that from the Earth's orbit.  So it sometimes makes sense to fit for one and not fit for the other.

Reply:
That's no problem; freeze one and leave the other free.

Reply:
Right.  But why is it an *error* if you don't set one?  In this case -- because of what I mentioned above -- I think that if one is unset, it should be assumed to be zero (which is what we do with basically all other unset params).

Reply:
Actually not. For various parameters, if they are needed and you haven't set them, you get an exception out of validate(). We aren't completely consistent about this; for some we do indeed default to treating them as zero (e.g. Taylor series terms).

Reply:
I think all of our "derivative" terms are like this.  And proper motion is a derivative term.  I think I'd really prefer to allow them to be set individually (and also EPS1DOT and EPS2DOT, which we've had discussions about before).  I haven't heard a good reason for why it should be an error, and people definitely do have good reasons for having just one set.

Reply:
To put it blunter:  there is nothing erroneous about setting a single proper motion component, so it shouldn't be an error.  If you are worried about a user not knowing what they are doing, maybe give a warning?

Reply:
It think it is somewhat common for just one PM component to be added, particularly many low dec pulsars might have a well determined PMRA but a poorly determined PMDEC, so the first step might be adding PMRA only.  I just loaded a model (NGC6440E.par.good) with no PM at all and `m.PMRA` is 0.0, not `None`.  So, I think testing for `None` will not be useful.  PM can't really be "UNSET". If you have astrometry you have a PM that is defaulted to 0.0, whether you like it or not, right?

Reply:
It is definitely common -- I do it all the time myself, as bulge globular clusters are near the ecliptic plane and so PMRA or PMELONG are each *way* more measureable than the other param.   In this case it doesn't matter that they are formally a vector given that it is the measurability that is the point.  And once again, it isn't an error to have just one.  It doesn't seem like a big deal that the tests/checks need to check for either of them to be set since (IMO) that's the right thing to do!

Reply:
PINT's handling of parameters not included in the par file is, as usual, various.

- If a parameter is given a default value upon construction (as, it seems, PMRA is), then if not present in the par file it will acquire that value just as if it had been set in the par file, and as a result it will be written out in output par files.
- If a parameter is not given a default value upon construction, then from Python the parameter's `.value` reads as None, it can be identified as unset, and code needs to cope with this. But it is not written to output par files. 
- If a parameter is part of a prefix family (e.g. F17) not present in the par file, it does not exist at all, and code needs to use `hasattr` to check whether it even exists before proceeding.

Checks of whether needed parameters exist in the par file are carried out during timing model creation, when the needed Components are picked, and also in the `validate()` function, and in some cases also in every function that needs to operate on the parameters.

None of these three behaviours seem to align with the usage pattern you describe - with PINT's current behaviour, the output par files will acquire PMRA and PMDEC, or PMELAT and PMELONG, whether both or one or none are set (I think this is a mistake). PINT could be adapted to treat missing PMRA (say) as zero but not add it to the par file, though this does require every use of PMRA from the python level to handle None appropriately.

(I should say that binary parameters, because they use the "standalone" wrappers, have different handling.)

In the short term, this code is clearly wrong because PMRA PMDEC PMELAT and PMELONG default to zero, so they will always exist and be set and written to output par files. But I think we should review how PINT handles parameters and default values. (The returning None for unset was triggering those numpy segfaults, some months ago.)

Reply:
Created issue #1124 to discuss this.

Reply:
Er, are you sure of that behaviour, Paul? The following tests both pass.

```python
par_basic_ecliptic = """
PSR J1234+5678
F0 1
PEPOCH 57000
ELAT 0
ELONG 0
DM 10
"""

par_basic_equatorial = """
PSR J1234+5678
F0 1
PEPOCH 57000
RAJ 10:23:47.67
DECJ 00:38:41.2
DM 10
"""

def test_pm_unset():
    m = get_model(StringIO(par_basic_ecliptic))
    assert m.PMELAT.quantity is None
    assert m.PMELONG.quantity is None
    assert m.POSEPOCH.quantity is None

def test_pm_unset_equatorial():
    m = get_model(StringIO(par_basic_equatorial))
    assert m.PMRA.quantity is None
    assert m.PMDEC.quantity is None
    assert m.POSEPOCH.quantity is None
```

Reply:
Here is what I tried (in the examples dir):
```
In [1]: import pint.models
In [2]: m = pint.models.get_model("NGC6440E.par.good")
/Users/paulr/src/PINT/src/pint/models/timing_model.py:344: UserWarning: PINT only supports 'T2CMETHOD IAU2000B'
  warn("PINT only supports 'T2CMETHOD IAU2000B'")

In [3]: m.PMRA
Out[3]: floatParameter( PMRA                0.0               (mas / yr) frozen=True)

In [4]: m.PMRA.quantity is None
Out[4]: False
```

Reply:
Oh, I see. I'm not doing that in this PR.  I see this PR has deleted the default value of 0.0 for PMRA.  It seems to me the defaulting PM to 0 makes sense.  Not sure what is gained by removing that default?

Reply:
As I said above, setting the default value to zero ensures that all output par files that are in equatorial coordinates will contain PMRA. I would prefer that the output par files resemble tidied normalized versions of the input par files rather than acquiring a bunch of values the user never specified. 

This is separate from treating unset parameters as zero, which the PR still does.

Reply:
That seems like a bug in the parfile writer:  Don't write out PMRA and PMDEC if the positions of the pulsars are in ecliptic coords (and vice versa).

Derivative values (IMO), of all types, should be de facto zero until explicitly set otherwise.  That is quite different from, say, M2, or SINI, or OM if ECC=0.  To do that in a nice way, we might need to have an "if set" boolean as you mentioned in #1124.

Reply:
And I see that you are saying that PMRA and PMDEC both get written out if you are using RA and DEC, as well.  But the same still applies -- don't write it out unless it is either non-zero or was explicitly set by the user (perhaps via your boolean idea).

Reply:
For me, I don't mind the behavior where if a par file is read in with only RAJ/DECJ, when written out it gets a PMRA/PMDEC that are both 0.0.  The AstrometryEquatorial always has proper motion as part of the model, so specifying them explicitly as zero in the output par doesn't really bother me.

Reply:
I think we agree on how unset values should be treated in computations - generally as zero, unless there is some other or no natural default. One question is when values that weren't set in the input get added to the output par file. Another question is how can we avoid nuisance and error in code that wants to use parameters that may be unset. Both are probably more helpfully discussed in #1124. For this one, since there seem to be par files in the wild that set only one of the PM parameters, I'd better change it to accept that.

Reply:
OK, put back the default value. We should do something about #1124 but this now behaves fine with only one proper motion parameter set.